File: 0005.Longest_Palindromic_Substring/Solution.java
class Solution {
    // zero characters? - ""
    // single character? - "a"
    // two characters? - "a"
    // spaces? unknown characters?
    // any algorihtm complexity/space requirements?
    public String longestPalindrome(String text) {
        if (isPalindrome(text)) {
            return text;
        }

        return findLongestExpandingCenter(text);
    }

    public String findLongestExpandingCenter(String text) {
        String longest = "";
        for (int index = 0; index < text.length(); index++) {
            int oddPalindromeLength = findPalindromeLength(text, index, index);
            int evenPalindromeLength = findPalindromeLength(text, index, index + 1);
            int maxLength = Math.max(oddPalindromeLength, evenPalindromeLength);
            if (maxLength > longest.length()) {
                longest = text.substring(index - ((maxLength - 1) / 2), index + (maxLength / 2) + 1);
            }
        }
        return longest;
    }

    private int findPalindromeLength(String text, int left, int right) {
        while (left >= 0 && right < text.length() && text.charAt(left) == text.charAt(right)) {
            left--;
            right++;
        }
        return right - left - 1;
    }
    
    public boolean isPalindrome(String text) {
        String reversed = new StringBuilder(text).reverse().toString();
        return text.equals(reversed);
    }
}

File: 0005.Longest_Palindromic_Substring/Solution.js
module.exports = { 
  //param A : string
  //return a strings
     longestPalindrome : function(text){
         // spaces/other characters?
         // empty array?
         // one character array?
         // 2 palindrome with same lengths?
         // is 'aa' also palindrome?
 
         // Solution 1
         // longestPalindrome = A[0]
         // foreach length
         //   foreach substring (i, length)
         //      is palindrome (from center to edges) or StringBuilder().reverse().equals()~ O(N)
         //      is longest
         //          longestPalindrome = A[0]
         // 
         // O(N^3)
         
         // Solution 2
         // longestPalindrome
         // foreach character
         //   while neigbourCharactersEqual and not out of range
         //      palindrome = leftNeigbour + palindrome + rightNeigbour
         //      palindromeLength += 2
         //   if palindromeLength > longestPalindrome
         //      longestPalindrome = palindrome
         //
         // O(N^2)
 
         // Solution 3
         // Map<char, list<index>>
         // foreach list<index> - substring, ifPalindrome
         // O(N^2)
         
         if (!text || text.length === 0) {
             return null;
         }
         
         function findPalindrome(base, left, right, text) {
             var palindrome = base;
             while (left >= 0 && right < text.length && text.charAt(left) === text.charAt(right)) {
                 palindrome = text.charAt(left) + palindrome + text.charAt(right);
                 left--;
                 right++;
             }
             return palindrome;
         }
 
         var longestPalindrome = text[0];
 
         for (var i = 0; i < text.length; ++i) {
             var oddPalindrome = findPalindrome(text.charAt(i), i - 1, i + 1, text);
             var evenPalindrome = findPalindrome('', i, i + 1, text);
 
             if (oddPalindrome.length > longestPalindrome.length) {
                 longestPalindrome = oddPalindrome;
             }
             if (evenPalindrome.length > longestPalindrome.length) {
                 longestPalindrome = evenPalindrome;
             }
         }
         
         return longestPalindrome;
     }
 };
 